suffuse/inode.go

```diff
@@ -0,0 +1,150 @@
+  "strings"
+  "sort"
+  "bazil.org/fuse"
+)
+
+/** Every directory inode has to have a way to obtain fresh
+ *  inodes to respond to e.g. Mkdir and Create calls. We could
+ *  reduce the footprint by splitting out Dir inodes from the
+ *  others, but for now every Inode carries the pointer.
+ */
+type Inode struct {
+  *InodeGen
+  AttrMap
+}
+
+func (x *Inode) AttrKeys()[]AttrKey {
+  keys := make([]AttrKey, 0, len(x.AttrMap))
+  for k := range x.AttrMap { keys = append(keys, k) }
+  return keys
+}
+func (x *Inode) ChildNames()[]Name {
+  ds := x.DirList()
+  buf := make([]string, 0)
+  for k := range ds { buf = append(buf, string(k)) }
+  sort.Strings(buf)
+  return Names(buf...)
+}
+
+func (x *Inode) String()string {
+  return Sprintf("%v[%v]", x.InodeType(), x.InodeNum())
+}
+func (x *Inode) TreeString()string {
+  return NewLines(expand(0, NoName, x)...).String()
+}
+
+func (x *Inode) SetAttr(k AttrKey, v AttrValue) {
+  x.AttrMap[k] = v
+}
+func (x *Inode) WithAttr(k AttrKey, v AttrValue)*Inode {
+  x.SetAttr(k, v)
+  return x
+}
+
+func (x *Inode) AttrOr(key AttrKey, alt AttrValue)(r AttrValue) {
+  r = x.AttrMap[key]
+  if r == nil { r = alt }
+  return
+}
+
+func (x *Inode) Text()string   { return string(x.Bytes())          }
+
+func (x *Inode) AddChildDir(name Name)(*Inode, error) {
+  if x.IsDir() {
+    child := x.NextDir()
+    x.DirList()[name] = child
+    return child, nil
+  }
+  return nil, NotADir()
+}
+func (x *Inode) AddChild(name Name, ino *Inode)error {
+  if x.IsDir() {
+    x.DirList()[name] = ino
+    return nil
+  }
+  return NotADir()
+}
+func (x *Inode) Child(name Name)*Inode {
+  if x.IsDir() {
+    return x.DirList()[name]
+  } else {
+    return nil
+  }
+}
+func (x *Inode) Dirents()[]fuse.Dirent {
+  res := make([]fuse.Dirent, 0)
+  for _, name := range x.ChildNames() {
+    child := x.Child(name)
+    res = append(res, child.FuseDirent(name))
+  }
+  return res
+}
+
+func (x *Inode) FuseDirent(name Name)fuse.Dirent {
+  return fuse.Dirent {
+    Inode: uint64(x.InodeNum()),
+    Type: x.InodeType().ToFuseType(),
+    Name: string(name),
+  }
+}
+
+func expand(level int, name Name, node *Inode)[]string {
+  indent := strings.Repeat(" ", level * 4)
+  line := Sprintf("%v[%v]%v", indent, node.InodeNum(), name)
+
+  switch node.InodeType() {
+    case InodeDir:
+      res := []string { line + "/" }
+      for _, name := range node.ChildNames() {
+        res = append(res, expand(level + 1, name, node.Child(name))...)
+      }
+      return res
+    case InodeLink:
+      return []string { Sprintf("%v -> %q", line, node.LinkTarget()) }
+    case InodeFile:
+      return []string { Sprintf("%v: %q", line, node.Text()) }
+    default:
+      return []string { line }
+  }
+}
+
+func (x *Inode) AddNodeMap(nodes map[Path]interface{})error {
+  for path, node := range nodes {
+    err := x.AddPath(path, node)
+    if err != nil { return err }
+  }
+  return nil
+}
+func (x *Inode) AddNode(segs []Name, node *Inode)error {
```

Comment:
I arrived at this function via `AddPath`. I wondered why `segs` was an array of `Name` instead of a path. It felt out of balance. Then I noticed the recursive call at the bottom. It seems this should not leak into the public API.

It makes me wonder if we should make a distinction between absolute paths and relative paths. The `AddPath` function ignores absolute paths and treats them as relative paths. I would like to hear your thoughts on this subject.